src/tests/utils/testing_tracer.rs

```diff
@@ -88,6 +97,8 @@ pub struct TestingTracer<const N: usize = 8, E: VmEncodingMode<N> = EncodingMode
     tracer_state: TracerState,
     /// stores pending messages that should be printed
     message_buffer: Option<String>,
+    /// Optional values pulled from assembly code denoting the storage refund and amount
+    test_refund_register_val: Option<Arc<Mutex<(StorageRefund, u32)>>>,
```

Comment:
And do we actually _need_ to use mutex at all?

Reply:
i tried a few different ways. another way we can try is by having the tracer take in a pointer to the storage oracle and call that function. right now we need it as the compiler complained without it

Reply:
i believe so, unless there's a better way to alter the data inside that field

Reply:
Using a mutex and passing it to the tracer feels wrong: 
1. At first I tried to understand where we implement multithreading :)
2. It's not good that the tracer has to know the implementation details

I can suggest next approach. Create separate struct with RefCell in `storage.rs` named something like "RefundController":

```rust
#[derive(Debug, Clone)]
pub struct RefundController {
    slot_refund: Rc<RefCell<(StorageRefund, u32)>>,
}

impl RefundController {
    pub fn new(slot_refund: Rc<RefCell<(StorageRefund, u32)>>) -> Self {
        Self { slot_refund }
    }

    pub fn set_storage_refund(&self, storage_refund_type: StorageRefund, refund_value: u32) {
        let new_value = if let StorageRefund::Warm = storage_refund_type {
            refund_value
        } else {
            0
        };
        *self.slot_refund.borrow_mut() = (storage_refund_type, new_value);
    }
}
```

And change `InMemoryCustomRefundStorage` implementation like:

```rust
impl InMemoryCustomRefundStorage {
    pub fn new() -> Self {
        Self {
            storage: InMemoryStorage::new(),
            slot_refund: Rc::new(RefCell::new((StorageRefund::Cold, 0u32))),
        }
    }

    pub fn create_refund_controller(&self) -> RefundController {
        RefundController::new(Rc::clone(&self.slot_refund))
    }
}
```

So it will be possible to create and move this structure to tracer:

```rust
// we are using TestingTracer to track prints and exceptions inside out_of_circuit_vm cycles
let mut out_of_circuit_tracer = TestingTracer::new(storage_impl.create_refund_controller());
```

As result, tracer will be able to just call `set_storage_refund` function

Reply:
Yeah, and `get_access_refund` will look like:

```rust
#[track_caller]
fn get_access_refund(
    &mut self,
    _monotonic_cycle_counter: u32,
    _partial_query: &LogQuery,
) -> StorageAccessRefund {
    let storage_refund = self.slot_refund.borrow();
    match storage_refund.0 {
        StorageRefund::Cold => dbg!(StorageAccessRefund::Cold),
        StorageRefund::Warm => dbg!(StorageAccessRefund::Warm {
            ergs: storage_refund.1
        }),
    }
}
```